test/server/listener_manager_impl_test.cc

```diff
@@ -1300,19 +1317,134 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, SingleFilterChainWithApplicationP
+  EXPECT_EQ(server_names.size(), 1);
+  EXPECT_EQ(server_names.front(), "server1.example.com");
+}
+
+TEST_F(ListenerManagerImplWithRealFiltersTest, SingleFilterChainWithSourceTypeMatch) {
+  const std::string yaml = TestEnvironment::substitute(R"EOF(
+    address:
+      socket_address: { address: 127.0.0.1, port_value: 1234 }
+    listener_filters:
+    - name: "envoy.listener.tls_inspector"
+      config: {}
+    filter_chains:
+    - filter_chain_match:
+        source_type: LOCAL
+      tls_context:
+        common_tls_context:
+          tls_certificates:
+            - certificate_chain: { filename: "{{ test_rundir }}/test/common/ssl/test_data/san_dns_cert.pem" }
+              private_key: { filename: "{{ test_rundir }}/test/common/ssl/test_data/san_dns_key.pem" }
+  )EOF",
+                                                       Network::Address::IpVersion::v4);
+
+  EXPECT_CALL(server_.random_, uuid());
+  EXPECT_CALL(listener_factory_, createListenSocket(_, _, true));
+  manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), "", true);
+  EXPECT_EQ(1U, manager_->listeners().size());
+
+  // TLS client without ALPN - no match.
+  auto filter_chain =
+      findFilterChain(1234, 3, "8.8.8.8", "127.0.0.1", true, "", true, "tls", true, {});
+  EXPECT_EQ(filter_chain, nullptr);
+
+  // TLS client with "http/1.1" ALPN - using 1st filter chain.
+  filter_chain = findFilterChain(1234, 3, "127.0.0.1", "127.0.0.1", true, "", true, "tls", true,
+                                 {"h2", "http/1.1"});
+  ASSERT_NE(filter_chain, nullptr);
+  EXPECT_TRUE(filter_chain->transportSocketFactory().implementsSecureTransport());
+  auto transport_socket = filter_chain->transportSocketFactory().createTransportSocket();
+  auto ssl_socket = dynamic_cast<Ssl::SslSocket*>(transport_socket.get());
+  auto server_names = ssl_socket->dnsSansLocalCertificate();
+  EXPECT_EQ(server_names.size(), 1);
+  EXPECT_EQ(server_names.front(), "server1.example.com");
+
+  // No filter chain defined for EXTERNAL match
+  filter_chain = findFilterChain(1234, 3, "4.4.4.4", "8.8.8.8", true, "", true, "tls", true, {});
+  EXPECT_EQ(filter_chain, nullptr);
+}
+
+TEST_F(ListenerManagerImplWithRealFiltersTest, MultipleFilterChainWithSourceTypeMatch) {
+  const std::string yaml = TestEnvironment::substitute(R"EOF(
+    address:
+      socket_address: { address: 127.0.0.1, port_value: 1234 }
+    listener_filters:
+    - name: "envoy.listener.tls_inspector"
+      config: {}
+    filter_chains:
+    - filter_chain_match:
+        source_type: LOCAL
+      tls_context:
+        common_tls_context:
+          tls_certificates:
+            - certificate_chain: { filename: "{{ test_rundir }}/test/common/ssl/test_data/san_dns_cert.pem" }
+              private_key: { filename: "{{ test_rundir }}/test/common/ssl/test_data/san_dns_key.pem" }
+    - filter_chain_match:
+        application_protocols: "http/1.1"
+        source_type: EXTERNAL
+      tls_context:
```

Comment:
Can you slim down these examples? We don't need TLS I think for example to validate filter chain selection, so this will be less boiler plate. Thanks.

Reply:
I don't know how to do that. I was just following whatever the rest of the `FilterChain` tests do.

Reply:
Just remove the TLS bits?

Reply:
The TLS checks are used to distinguish between the different filters when one fo them is matched. I can remove these but then we'll have no means to tell if the intended filter match was found.

Reply:
I see, but TBH, this is an overally complicated way of achieving this. TLS has nothing to do with the property being tested and you should be able to distinguish filter chains in other ways. That said, probably not worth changing for this PR unless you are eager to clean up the other places we do this.

Reply:
I agree it is too complicated. I was trying to figure out the semantics of `listener.Filter` and how to use it to identify the matched filter, but failed to do so. Maybe if we can have an issue to improve this tests with a suggestion what could be used instead of TLS?